File: back/runtime/adoption/adoption.go
// Package adoption is the AIDOS Runtime AdoptionStage ladder (step S47): the pure,
// read-only function that names the SMALLEST RATCHET THAT CLICKS NEXT for an adopter
// installing KRD progressively (KRD §82.5 — "install the smallest ratchet that
// clicks, then extend by scars"). It COMPUTES a ladder; it never installs, mutates,
// or writes truth.
//
// FIVE DECLARED TIERS, NEVER INVENTED (CLAUDE.md honesty rule). The ladder is EXACTLY
// five tiers T0..T4 — mapped from KRD §82.5's stage0..stage5, REUSED not re-coined:
//
//   - T0 — existing tests + mutation (the floor every project already has).
//   - T1 — one KRD cell (the first verticale: idea → mirror → goal → green).
//   - T2 — kernel + mirror (the truth-store + its bicephalous proof plane).
//   - T3 — ContextGraph + Memory (the branch-aware context router + the brain).
//   - T4 — evolve + QualityDiversity (the medium loop in its EvolutionSandbox).
//
// Each tier carries a DECLARED requires/grants over Capabilities. A tier is
// SATISFIABLE iff every capability it requires is live in the passed-in view.
//
// THREE LOAD-BEARING GATING FACTS (the done criteria, KRD-pinned):
//   - T1 does NOT require QualityDiversity — QD is an ADVANCED capability (KRD §82.6,
//     not in the mandatory_minimum); asserting it at T1 is a monster.
//   - T2 REQUIRES a live RealityMirror — the external-world mirror (KRD Livre XX) must
//     ground a cell before it may enter the catastrophic tier.
//   - T4 REQUIRES a live EvolutionSandbox — the /evolve quarantine (KRD §66.1) must
//     exist before evolution may be turned on.
//
// DETERMINISM-FIRST (CLAUDE.md §6/§8). Plan is a pure, total function of capabilities
// alone — no DB, no I/O, no time.Now(), no RNG: same inputs ⇒ same plan. Parsing /
// gating / smallest-next-tier selection is code, never an LLM. The reproducibility
// property mirror pins it.
//
// THE WALL (CLAUDE.md §2). This package is PURE and READ-ONLY over a capability view
// (passed in). It writes NOTHING. Plan returns the report, never a boolean it then
// satisfies (anti-Goodhart). Turning on a capability (a RealityMirror, an
// EvolutionSandbox, QD) is a PRIOR/OTHER step's truth, CONSUMED here, never produced.
package adoption

// Capability is one capability an adopter may have live in the truth-store. These are
// CONSUMED facts (live or not) — this package never builds or turns one on. The set is
// declared (the gating predicates reference exactly these); the rapid property pins
// that no tier requires a capability outside the declared map.
type Capability string

const (
	// CapTests — an existing test suite (the floor; T0's grant).
	CapTests Capability = "tests"
	// CapMutation — mutation testing over those tests (T0's grant).
	CapMutation Capability = "mutation"
	// CapOneKRDCell — one full KRD cell wired (idea → mirror → goal → green); T1's grant.
	CapOneKRDCell Capability = "one_krd_cell"
	// CapKernel — the kernel truth-store (DSL ASTs in Postgres); T2's grant.
	CapKernel Capability = "kernel"
	// CapMirror — the mirrors plane (the bicephalous proof); T2's grant.
	CapMirror Capability = "mirror"
	// CapRealityMirrorLive — a LIVE RealityMirror (the external-world mirror, KRD
	// Livre XX). T2's load-bearing requirement: no catastrophic tier without reality
	// grounding. CONSUMED, never built here.
	CapRealityMirrorLive Capability = "reality_mirror_live"
	// CapContextGraph — the branch-aware ContextGraph/router; T3's grant.
	CapContextGraph Capability = "context_graph"
	// CapMemory — the brain (pgvector memory); T3's grant.
	CapMemory Capability = "memory"
	// CapEvolutionSandbox — a live EvolutionSandbox (the /evolve quarantine, KRD
	// §66.1). T4's load-bearing requirement: no evolution before its sandbox exists.
	// CONSUMED, never built here.
	CapEvolutionSandbox Capability = "evolution_sandbox"
	// CapEvolve — the /evolve medium loop turned on; T4's grant.
	CapEvolve Capability = "evolve"
	// CapQualityDiversity — QD search. KRD §82.6: ADVANCED, NOT in the mandatory
	// minimum — it is T4's grant and is NEVER a T1 requirement (the monster guard).
	CapQualityDiversity Capability = "quality_diversity"
)

// AdoptionStage is the typed ladder tier — EXACTLY five, declared, never invented.
// Mapped from KRD §82.5's stage0..stage5 (REUSED, not re-coined). The rapid property
// pins AdoptionStage ∈ { T0,T1,T2,T3,T4 }.
type AdoptionStage string

const (
	// T0 — existing tests + mutation (KRD §82.5 stage0).
	T0 AdoptionStage = "T0"
	// T1 — one KRD cell (stage1). Does NOT require QualityDiversity (§82.6).
	T1 AdoptionStage = "T1"
	// T2 — kernel + mirror (stage2). Requires a live RealityMirror (Livre XX).
	T2 AdoptionStage = "T2"
	// T3 — ContextGraph + Memory (stage3).
	T3 AdoptionStage = "T3"
	// T4 — evolve + QualityDiversity (stage4). Requires a live EvolutionSandbox (§66.1).
	T4 AdoptionStage = "T4"
)

// Stages returns the five tiers in ladder order (T0 first). Used by Plan and by the
// Workbench projection so the set of tiers is never invented.
func Stages() []AdoptionStage { return []AdoptionStage{T0, T1, T2, T3, T4} }

// stageRank orders the tiers for the monotone ladder (lower tier first). A tier is
// proposable only once every lower tier is satisfied.
var stageRank = map[AdoptionStage]int{T0: 0, T1: 1, T2: 2, T3: 3, T4: 4}

// IsStage reports whether s is one of the five declared tiers.
func IsStage(s AdoptionStage) bool { _, ok := stageRank[s]; return ok }

// tierSpec is the DECLARED requires/grants of one tier (above the line — declared,
// never learned). requires is the closed set a tier needs live to be satisfiable;
// grants is what installing the tier turns on. The three load-bearing facts live in
// these declarations: T1.requires has NO QualityDiversity, T2.requires has the
// RealityMirror, T4.requires has the EvolutionSandbox.
type tierSpec struct {
	requires []Capability
	grants   []Capability
}

// specs is the declared ladder. The mapping is KRD §82.5's stage0..stage5 — REUSED.
// Each tier requires its OWN new capabilities (the cumulative lower-tier requirements
// are enforced by the monotone-ladder rule in Plan, not duplicated here).
var specs = map[AdoptionStage]tierSpec{
	// T0 — the floor: existing tests + mutation. Requires nothing (every project has it).
	T0: {requires: nil, grants: []Capability{CapTests, CapMutation}},
	// T1 — one KRD cell. Requires the floor + one wired cell. Does NOT require QD (§82.6).
	T1: {requires: []Capability{CapTests, CapMutation, CapOneKRDCell}, grants: []Capability{CapOneKRDCell}},
	// T2 — kernel + mirror + a LIVE RealityMirror (Livre XX — the load-bearing fact).
	T2: {requires: []Capability{CapKernel, CapMirror, CapRealityMirrorLive}, grants: []Capability{CapKernel, CapMirror}},
	// T3 — ContextGraph + Memory.
	T3: {requires: []Capability{CapContextGraph, CapMemory}, grants: []Capability{CapContextGraph, CapMemory}},
	// T4 — evolve + QD, gated on a live EvolutionSandbox (§66.1 — the load-bearing fact).
	T4: {requires: []Capability{CapEvolutionSandbox, CapEvolve, CapQualityDiversity}, grants: []Capability{CapEvolve, CapQualityDiversity}},
}

// Requires returns the declared requirements of a tier (defensive copy — the input
// declaration is never exposed for mutation). Used by the panel and the property test
// (T1.requires never contains QualityDiversity).
func Requires(s AdoptionStage) []Capability {
	r := specs[s].requires
	out := make([]Capability, len(r))
	copy(out, r)
	return out
}

// Grants returns the declared grants of a tier (defensive copy).
func Grants(s AdoptionStage) []Capability {
	g := specs[s].grants
	out := make([]Capability, len(g))
	copy(out, g)
	return out
}

// Gap is one missing capability blocking a tier — it names the EXACT capability the
// view lacks (never an invented requirement) and the tier it blocks, in the ubiquitous
// language. The honesty rule: a Gap always traces to a declared requires entry.
type Gap struct {
	Stage   AdoptionStage `json:"stage"`
	Missing Capability    `json:"missing"`
	Reason  string        `json:"reason"`
}

// TierStatus is the computed status of one tier against the capability view: whether it
// is satisfiable, and the Gaps blocking it when not. Read-only; computed, never declared.
type TierStatus struct {
	Stage       AdoptionStage `json:"stage"`
	Requires    []Capability  `json:"requires"`
	Grants      []Capability  `json:"grants"`
	Satisfiable bool          `json:"satisfiable"`
	Gaps        []Gap         `json:"gaps,omitempty"`
}

// AdoptionPlan is the typed ladder report: every tier's status, the CURRENT tier (the
// highest contiguously-satisfied tier from the floor), and the NEXT smallest installable
// tier (the smallest ratchet that clicks) with its blocking gaps. Computed purely from
// the capability view; never declared.
type AdoptionPlan struct {
	Tiers   []TierStatus  `json:"tiers"`
	Current AdoptionStage `json:"current"`
	// Next is the smallest unsatisfied tier whose lower tiers are all satisfied — the
	// next ratchet that clicks. Empty when every tier is already satisfied.
	Next     AdoptionStage `json:"next,omitempty"`
	NextGaps []Gap         `json:"next_gaps,omitempty"`
	// AllSatisfied is true when every tier T0..T4 is satisfiable in the view.
	AllSatisfied bool `json:"all_satisfied"`
}

// reasonFor names a missing capability in the ubiquitous language. The three
// load-bearing facts get a KRD-pinned reason; the rest a uniform one.
func reasonFor(s AdoptionStage, c Capability) string {
	switch c {
	case CapRealityMirrorLive:
		return "bloqué jusqu'à ce qu'un RealityMirror soit vivant (le miroir du monde extérieur, KRD Livre XX — pas de palier catastrophique sans ancrage dans le réel)"
	case CapEvolutionSandbox:
		return "bloqué jusqu'à ce que l'EvolutionSandbox existe (la quarantaine de /evolve, KRD §66.1 — pas d'évolution avant son bac à sable)"
	default:
		return string(s) + " requiert la capacité « " + string(c) + " » : absente de la vue (CONSOMMÉE, jamais inventée)"
	}
}

// satisfiable reports whether every capability the tier requires is live in have, and
// returns the ordered Gaps (one per missing requirement) when not. Pure over the view.
func (ts tierSpec) status(s AdoptionStage, have map[Capability]bool) TierStatus {
	var gaps []Gap
	for _, c := range ts.requires {
		if !have[c] {
			gaps = append(gaps, Gap{Stage: s, Missing: c, Reason: reasonFor(s, c)})
		}
	}
	return TierStatus{
		Stage:       s,
		Requires:    append([]Capability(nil), ts.requires...),
		Grants:      append([]Capability(nil), ts.grants...),
		Satisfiable: len(gaps) == 0,
		Gaps:        gaps,
	}
}

// Plan computes the AdoptionPlan from the live capability view (the smallest ratchet
// that clicks next). Pure and total — no DB, no I/O, no time.Now(), no RNG; same
// capabilities ⇒ same plan (the determinism property). It NEVER installs, mutates the
// input, or writes truth (the wall). Malformed/unknown capabilities in the view are
// ignored (never panic) — only the declared ones gate.
//
// The CURRENT tier is the highest tier such that every tier from T0 up to it is
// satisfiable (the contiguous floor). The NEXT tier is the smallest unsatisfied tier
// whose lower tiers are all satisfied — the monotone-ladder rule (no jumping to T2+
// while T1 is unmet).
func Plan(capabilities []Capability) AdoptionPlan {
	have := make(map[Capability]bool, len(capabilities))
	for _, c := range capabilities {
		have[c] = true // unknown capabilities are simply present-but-ungating.
	}

	stages := Stages()
	tiers := make([]TierStatus, 0, len(stages))
	for _, s := range stages {
		tiers = append(tiers, specs[s].status(s, have))
	}

	plan := AdoptionPlan{Tiers: tiers}

	// Walk the ladder from the floor: current = the highest contiguously-satisfied
	// tier; next = the first unsatisfied tier reached (lower tiers all satisfied).
	plan.Current = ""
	allSat := true
	nextSet := false
	for i, ts := range tiers {
		if ts.Satisfiable {
			if !nextSet { // still on the contiguous satisfied floor → advance current.
				plan.Current = stages[i]
			}
			continue
		}
		allSat = false
		if !nextSet {
			plan.Next = ts.Stage
			plan.NextGaps = append([]Gap(nil), ts.Gaps...)
			nextSet = true
		}
	}
	plan.AllSatisfied = allSat
	return plan
}

File: front/web/lib/adoption.ts
// Determinism-first twin of back/runtime/adoption.Plan + back/runtime/adoption/release
// .Assemble (KRD §S47). PURE functions — no clock read (now is passed), no rng, no I/O
// — so /adoption can re-run the ladder + the pack assembly on a selected scenario and
// render the EXACT AdoptionPlan + ReleasePack the Go package computes. The five tiers,
// the three load-bearing gating facts and the assemble-only/no-author rule mirror the
// Go package; fast-check (lib/adoption.test.ts) pins the twin's invariants.
//
// THE WALL (CLAUDE.md §2/§8): the twin COMPUTES a ladder + ASSEMBLES an inventory — it
// installs nothing, ships nothing, writes no truth, opens no ChangeSet. Turning on a
// capability goes through the door `idea → mirror → /goal → human approval`. The
// recorded pack lives in fitness.release_pack (SELECT-only to the agent); this twin
// renders it.

// The eleven DECLARED capabilities — CONSUMED facts (live or not), never built here.
export type Capability =
	| "tests"
	| "mutation"
	| "one_krd_cell"
	| "kernel"
	| "mirror"
	| "reality_mirror_live"
	| "context_graph"
	| "memory"
	| "evolution_sandbox"
	| "evolve"
	| "quality_diversity";

// The five DECLARED tiers — exactly these (KRD §82.5 stage0..stage5), never invented.
export type AdoptionStage = "T0" | "T1" | "T2" | "T3" | "T4";

export const STAGES: readonly AdoptionStage[] = [
	"T0",
	"T1",
	"T2",
	"T3",
	"T4",
] as const;

const STAGE_RANK: Record<AdoptionStage, number> = {
	T0: 0,
	T1: 1,
	T2: 2,
	T3: 3,
	T4: 4,
};

export function isStage(s: string): s is AdoptionStage {
	return s in STAGE_RANK;
}

// The single door any capability install requires — /release never bypasses it.
export const THE_DOOR = "idea → mirror → /goal → human approval";

interface TierSpec {
	requires: Capability[];
	grants: Capability[];
}

// The declared ladder — REUSED from KRD §82.5, mirrors the Go `specs`. T1.requires has
// NO quality_diversity (§82.6); T2.requires has reality_mirror_live (Livre XX);
// T4.requires has evolution_sandbox (§66.1).
const SPECS: Record<AdoptionStage, TierSpec> = {
	T0: { requires: [], grants: ["tests", "mutation"] },
	T1: {
		requires: ["tests", "mutation", "one_krd_cell"],
		grants: ["one_krd_cell"],
	},
	T2: {
		requires: ["kernel", "mirror", "reality_mirror_live"],
		grants: ["kernel", "mirror"],
	},
	T3: {
		requires: ["context_graph", "memory"],
		grants: ["context_graph", "memory"],
	},
	T4: {
		requires: ["evolution_sandbox", "evolve", "quality_diversity"],
		grants: ["evolve", "quality_diversity"],
	},
};

export function requires(s: AdoptionStage): Capability[] {
	return [...SPECS[s].requires];
}

export function grants(s: AdoptionStage): Capability[] {
	return [...SPECS[s].grants];
}

export interface Gap {
	stage: AdoptionStage;
	missing: Capability;
	reason: string;
}

export interface TierStatus {
	stage: AdoptionStage;
	requires: Capability[];
	grants: Capability[];
	satisfiable: boolean;
	gaps: Gap[];
}

export interface AdoptionPlan {
	tiers: TierStatus[];
	current: AdoptionStage | "";
	next: AdoptionStage | "";
	nextGaps: Gap[];
	allSatisfied: boolean;
}

function reasonFor(s: AdoptionStage, c: Capability): string {
	switch (c) {
		case "reality_mirror_live":
			return "bloqué jusqu'à ce qu'un RealityMirror soit vivant (le miroir du monde extérieur, KRD Livre XX — pas de palier catastrophique sans ancrage dans le réel)";
		case "evolution_sandbox":
			return "bloqué jusqu'à ce que l'EvolutionSandbox existe (la quarantaine de /evolve, KRD §66.1 — pas d'évolution avant son bac à sable)";
		default:
			return `${s} requiert la capacité « ${c} » : absente de la vue (CONSOMMÉE, jamais inventée)`;
	}
}

// plan computes the AdoptionPlan from the live capability view (the smallest ratchet
// that clicks next). Pure; never installs; never mutates the input. The monotone-ladder
// rule: current = the contiguous satisfied floor, next = the first unsatisfied tier.
export function plan(capabilities: Capability[]): AdoptionPlan {
	const have = new Set(capabilities);
	const tiers: TierStatus[] = STAGES.map((s) => {
		const spec = SPECS[s];
		const gaps: Gap[] = [];
		for (const c of spec.requires) {
			if (!have.has(c)) {
				gaps.push({ stage: s, missing: c, reason: reasonFor(s, c) });
			}
		}
		return {
			stage: s,
			requires: [...spec.requires],
			grants: [...spec.grants],
			satisfiable: gaps.length === 0,
			gaps,
		};
	});

	let current: AdoptionStage | "" = "";
	let next: AdoptionStage | "" = "";
	let nextGaps: Gap[] = [];
	let allSatisfied = true;
	let nextSet = false;
	for (const ts of tiers) {
		if (ts.satisfiable) {
			if (!nextSet) current = ts.stage;
			continue;
		}
		allSatisfied = false;
		if (!nextSet) {
			next = ts.stage;
			nextGaps = [...ts.gaps];
			nextSet = true;
		}
	}
	return { tiers, current, next, nextGaps, allSatisfied };
}

// ─── release.Assemble twin ───────────────────────────────────────────────────

export interface CLICommand {
	name: string;
	summary?: string;
}
export interface Route {
	path: string;
	title?: string;
}
export interface DemoRef {
	ref?: string;
	title?: string;
}
export interface DocRef {
	slug: string;
	title?: string;
}
export interface MirrorRef {
	id: string;
	reflects?: string;
	testKind?: string;
}
export interface ChangeEntry {
	ref: string;
	summary?: string;
}
export interface Limit {
	ref: string;
	description?: string;
}

export interface View {
	cliCommands: CLICommand[];
	workbenchRoutes: Route[];
	demo: DemoRef;
	docs: DocRef[];
	mirrors: MirrorRef[];
	changesets: ChangeEntry[];
	declaredLimits: Limit[];
	kernelHead: string;
}

export interface ReleasePack {
	id: string;
	cliSurface: CLICommand[];
	workbenchRoutes: Route[];
	demoCell: DemoRef;
	docsIndex: DocRef[];
	testInventory: MirrorRef[];
	changelog: ChangeEntry[];
	knownLimits: Limit[];
	adoptionPlan: AdoptionPlan;
	assembledAt: number;
	kernelHead: string;
}

// A stable, deterministic synthetic id for a pack (the Go package content-addresses
// with Hash(Canonicalize(body))). The twin uses a length-fingerprint synthetic id —
// enough for the panel's keys + the assemble-only proof (the panel renders the recorded
// pack's real hash when wired to fitness.release_pack).
function syntheticId(p: Omit<ReleasePack, "id" | "assembledAt">): string {
	return `pack:${p.cliSurface.length}c-${p.workbenchRoutes.length}r-${p.testInventory.length}m-${p.changelog.length}cl-${p.knownLimits.length}kl-${p.kernelHead}`;
}

// assemble inventories the live View into a ReleasePack paired with the adoption ladder.
// Pure; AUTHORS NOTHING (every field is the View's content); installs/ships nothing;
// never mutates the input. An empty View yields an empty-but-valid pack.
export function assemble(
	view: View,
	capabilities: Capability[],
	now: number,
): ReleasePack {
	const cli = [...view.cliCommands].sort((a, b) =>
		a.name < b.name ? -1 : a.name > b.name ? 1 : 0,
	);
	const routes = [...view.workbenchRoutes].sort((a, b) =>
		a.path < b.path ? -1 : a.path > b.path ? 1 : 0,
	);
	const docs = [...view.docs].sort((a, b) =>
		a.slug < b.slug ? -1 : a.slug > b.slug ? 1 : 0,
	);
	const mirrors = [...view.mirrors].sort((a, b) =>
		a.id < b.id ? -1 : a.id > b.id ? 1 : 0,
	);
	const changelog = [...view.changesets].sort((a, b) =>
		a.ref < b.ref ? -1 : a.ref > b.ref ? 1 : 0,
	);
	const limits = [...view.declaredLimits].sort((a, b) =>
		a.ref < b.ref ? -1 : a.ref > b.ref ? 1 : 0,
	);
	const adoptionPlan = plan(capabilities);

	const partial = {
		cliSurface: cli,
		workbenchRoutes: routes,
		demoCell: view.demo,
		docsIndex: docs,
		testInventory: mirrors,
		changelog,
		knownLimits: limits,
		adoptionPlan,
		kernelHead: view.kernelHead,
	};
	return { ...partial, id: syntheticId(partial), assembledAt: now };
}
